keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,582 @@
+
+In both scenarios, the *Memory Manager* employs hint generation protocol to yield the most suitable NUMA affinity for a pod, and it feeds the central manager (*Topology Manager*) with those affinity hints. Moreover, *Memory Manager* ensures that the memory, which a pod requests, is allocated from a minimum number of NUMA nodes.
+
+The up-to-date implementation of the *Memory Manager* is available through the link in [Implementation History](##implementation-history) section. Technically, the *single-NUMA* strategy is a special case of the *multi-NUMA* strategy, and thus, no separate implementation was developed for them. Still, the proposal differentiates between the *single-NUMA* and the *multi-NUMA* strategies. First of all, it allows for a number of important use cases or situations to be discussed and contrasted.       
+
+## Motivation
+
+### Goals
+
+- Offer guaranteed memory (and hugepages) allocation over a minimum number of NUMA nodes for containers (within a pod). 
+
+- Guaranteeing the affinity of memory and hugepages to the same NUMA node for the whole group of containers (within a pod). This is a long-term goal which will be achieved along with PR [#1752](https://github.com/kubernetes/enhancements/pull/1752) and the implementation of `hintprovider.GetPodLevelTopologyHints()` API in the *Memory Manager*.  
+
+### Non-Goals
+
+- Both the upgrade of K8S scheduler and further extension of the official pod specification are out of scope.
+
+- This proposal confines to the family of Linux operating systems.
+
+## Proposal
+
+The *Memory Manager* is a new component of _kubelet_ ecosystem proposed to enable single-NUMA and multi-NUMA guaranteed memory allocation. Besides:
+- *Memory Manager* is also a new hint provider for _Topology Manager_;
+- The hints are intended to indicate preferred memory (and hugepages) affinity which pins the memory for a container either to a single or a group of NUMA nodes;
+- By design, *Memory Manager* offers guaranteed memory allocation for pods in Guaranteed QoS class.
+
+The next subsection introduces to the proposal by outlining its design. 
+
+### Design Overview
+
+![alt text](./mm_tm_diagram.png)
+
+Once *kubelet* requests a guaranteed QoS pod admission, as shown in the figure above, *Topology Manager* queries *Memory Manager* about the preferred NUMA affinity for memory and hugepages, for all containers in a pod. For each container in a pod, *Memory Manager* calculates the affinity using its internal database, namely the `Node Map`. The `Node Map` is an object which is responsible for tracking the utilization of memory (and hugepages) for all containers in Guaranteed QoS class. 
```

Comment:
Does it make sense to talk about hugepages separate from memory?  You could just track page size and number of pages and it would be generic.

Reply:
We use exactly the same logic for hugepages calculation and hugepages NUMA node allocation affected as well by `cpuset.mems`, so I do not think we need to talk separately about memory and hugepages.